website/blog/2024-03-18-deep-learning-part-2/index.md

```diff
@@ -0,0 +1,230 @@
+--num_gpus='{num_gpus}' \
+--module databricks_llm.fine_tune \
+--final_model_output_path='{dbfs_output_location}' \
+--output_dir="/local_disk0/output" \
+--dataset={dataset} \
+--model={pretrained_name_or_path} \
+--tokenizer={pretrained_name_or_path} \
+--use_lora=false \
+--use_4bit=false \
+--deepspeed_config="ds_configs/ds_zero_3_cpu_offloading.json" \
+--fp16=false \
+--bf16=true \
+--per_device_train_batch_size=16 \
+--per_device_eval_batch_size=48 \
+--gradient_checkpointing=true \
+--gradient_accumulation_steps=1 \
+--learning_rate=1e-6 \
+--lr_scheduler_type="cosine" \
+--warmup_steps=50 \
+--evaluation_strategy="steps" \
+--save_strategy="steps" \
+--save_steps=100 \
+--num_train_epochs=3\
+--run_name="{pretrained_name_or_path}-blog-run"
+``````
+See code details [here](https://github.com/puneet-jain159/fine_tuning_llm_with_mlflow_blog/blob/495c156f73435996706c42bfda4bb7519dc21dab/notebooks/run_deepspeed_fine_tuning.py#L117)
+
+In the above code block we are doing the following: Logging system metrics with MLFlow ensures that CPU stats, GPU stats, memory usage, network traffic, and disk usage during the training process is logged. Particularly, the ability to log and track GPU utilization during a training run is critical and thus being able to log it intermittently is particularly useful. The training run is parametrized such that the model is trained for 3 epochs, with checkpoints saved and metrics logged every 100 training steps. The run is given an appropriate name based on the path of the pretrained model.
+
+The diagram below is a visualization of how finetuning and MLflow fit together.
+
+![](finetune_with_MLflow.png)
+
+Using MLflow while fine tuning ML model in Databricks
+
+
+# The Importance of Logging
+
+The integration of the Hugging Face trainer callback with MLFlow plays a pivotal role in this ecosystem. It allows for comprehensive tracking and logging of metrics, parameters, and artifacts during model finetuning. This logging can be configured flexibly – at every epoch, after a set number of epochs, or steps. Importantly, it can cover the entire model checkpoint or be limited to just the adapter when using [PEFT (Parameter-Efficient Fine-Tuning)](https://huggingface.co/docs/peft/index) methods like [QLoRA](https://arxiv.org/abs/2305.14314). We have highlighted the additional code necessary for MLflow to log every epoch and checkpoint.
+
+
+## Configuring Hugging Face Trainer Callback for MLflow
+
+```python
+class CustomMLflowCallback(MLflowCallback):
+    """
+    A [`TrainerCallback`] that sends the logs to [MLflow](https://www.mlflow.org/). Can be disabled by setting
+    environment variable `DISABLE_MLFLOW_INTEGRATION = TRUE`.
+    """
+
+  def on_save(self, args, state, control,model,tokenizer, **kwargs):
+        if self._initialized and state.is_world_process_zero and self._log_artifacts:
+           ckpt_dir = f"checkpoint-{state.global_step}"
+  	 self._ml_flow.transformers.log_model(
```

Comment:
What is the reason for patching log_model? 
You shouldn't split the naming like this of `mlflow` to `_ml_flow`. `_mlflow` would be sufficient. 
I'm not sure what the intention of having an object instance reference to mlflow is, though. Why not set an object attr that indicates whether mlflow is installed and importable and adjust logic within the instance methods to handle calls after wrapping the import in a try/catch. 
I'm also a bit reticent in showing users an example of using PEFT that defines manually overriding pip_requirements (an anti-pattern). Is the reason for not using the standard flavor purely so that you can do checkpointing during training?
If the intention of the blog is to show off auto-checkpointing, it might be a better idea to use a torch or TF example that has this directly supported with standard public APIs. Transformers doesn't yet have this functionality (namely due to no one asking for it) and I feel like this example is entirely far too complex for people to read through.